firebase-firestore/ktx/src/main/kotlin/com/google/firebase/firestore/ktx/serialization/FirestoreMapDecoder.kt

```diff
@@ -0,0 +1,244 @@
+            }
+            else -> {
+                throw Exception("Incorrect format of nested data provided: <$innerCompositeObject>")
+            }
+        }
+    }
+
+    /**
+     * Returns the nested structured object that needs to be decoded. The nested structured object
+     * could be either a nested [Serializable] object or a nested List.
+     */
+    private fun getCompositeObject(elementIndex: Int): Any? {
+        return when (elementIndex) {
+            START_INDEX ->
+                nestedObject // the custom object itself is the first structured obj to decode
+            else -> decodeValueList[elementIndex - 1]
+        }
+    }
+}
+
+/**
+ * The entry point of Firestore Kotlin deserialization process. It decodes a nested map of Firestore
+ * supported types to a [Serializable] Kotlin object.
+ *
+ * For a [Serializable] object, at compile time, a serializer will be generated by the Kotlin
+ * serialization compiler plugin (or a custom serializer can be manually passed in). The structure
+ * information of the [Serializable] object will be recorded inside of the serializer’s descriptor
+ * (i.e. the name/type of each property to be encoded, the annotations on each property).
+ *
+ * Based on the descriptor’s information, during the decoding process, a nested map will be send to
+ * the decoder. Decoder will loop through all the key-value pairs in the nested map, use the key to
+ * find the target field from the [SerialDescriptor], then assign the value to the field.
+ *
+ * @param nestedObject The nested map that that needs to be decoded to a @[Serializable] object.
+ * @param docRef The [DocumentReference] where this nested map is obtained from.
+ */
+private class FirestoreMapDecoder(nestedObject: Map<String, Any?>, docRef: DocumentReference) :
+    FirestoreAbstractDecoder(nestedObject, docRef) {
+
+    private val nestedMap = nestedObject
+    private val decodeNameList: List<String>
+    override val decodeValueList: List<Any?>
+
+    /**
+     * Returns a list of keys and values that need to be decoded as fields of the [Serializable]
+     * object.
+     */
+    init {
+        val (decodeNameList, decodeValueList) = nestedObject.toList().unzip()
+        this.decodeNameList = decodeNameList
+        this.decodeValueList = decodeValueList
+    }
+
+    /**
+     * Returns the index of the element to be decoded. Index represents a position of the current
+     * element in the serial descriptor element that can be found with
+     * [SerialDescriptor.getElementIndex]. Additional to the element index, this method also returns
+     * [CompositeDecoder.DECODE_DONE] to indicate the decoding process is finished, and returns
+     * [CompositeDecoder.UNKNOWN_NAME] to indicate that the element to be decode is not in
+     * [descriptor].
+     *
+     * @param descriptor The [SerialDescriptor] of the [Serializable] object.
+     */
+    override fun decodeElementIndex(descriptor: SerialDescriptor): Int {
```

Comment:
no need to prefix all with `decode`

Reply:
This is the method name I have to override from the parent class `CompositeDecoder.kt`
https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-core/kotlinx.serialization.encoding/-composite-decoder/decode-element-index.html

Reply:
I meant the variables inside :-)